libr/crypto/p/crypto_serpent_algo.c

```diff
@@ -77,7 +77,7 @@ void apply_IP(ut32 in[DW_BY_BLOCK], ut32 out[DW_BY_BLOCK]) {
 	for (i = 0; i < DW_BY_BLOCK*32; i++) {
 		index = IPTable[i];
 		out[i/32] ^= (-(ut32)get_bit (index%32, in[index/32]) ^ out[i/32])
-			& (1 << i);
+			& (1 << (i & 0x1f));
```

Comment:
Wouldn't it be better to use `i % 32`?

Reply:
I don't know whether there's a compiler that won't do this conversion, so I used &0x1f directly. Just checked the assembly on rpi4 and yes, %32 is treated as &31 properly. Is there a readability concern?

Reply:
this is a shift left, this means that using % will result on wrong behaviour because it will be using a rotated value instead of a shifted one.

The variable `i` goes from 0 to DW_BY_BLOCK * 32, which is 4.. this is.. this code will try to shift from 0 to 128 bits. this is a clear UB. and for me it seems like using % will result in better behaviour. but its different than the current one. I'm not sure if we have tests for this or if you tried to compare this with other implementations of serpent. but if we are just xoring the buffer aganist 0s.. its probably better to break before doing useless ops..

Can you add a test and compare the result with another implementation. https://github.com/sysopfb/pyserpent this one seems like a good candidate for messing up and comparing outputs.